python/mlc_llm/quantization/per_tensor_quantization.py

```diff
@@ -357,10 +410,43 @@ def forward(self, x: nn.Tensor) -> nn.Tensor:  # pylint: disable=invalid-name
             The output tensor for the per-tensor quantized linear layer.
         """
         # Note: Use calibration scale when calibration is enabled
-        x = x.astype(self.config.activation_dtype)
-        if self.config.weight_dtype == self.config.storage_dtype:
+        if self.config.calibration_mode == "inference":
+            if self.q_calibration_scale:
+                x /= self.q_calibration_scale
+            x_q = x.astype(self.config.activation_dtype)
+            x_scale = self.q_calibration_scale
+        elif self.config.calibration_mode == "max":
+            _, x_scale = self.config.quantize_float8(  # type: ignore
+                x,
+                quantize_dtype=self.config.activation_dtype,
+                storage_dtype=self.config.storage_dtype,
+            )
+            if self.config.tensor_parallel_shards > 1:
+                x_scale = nn.ccl_allreduce(x_scale, "max")
+            x_scale = nn.extern(
+                "mlc_llm.calibration_observer",
+                [f"{self.name}.q_calibration_scale", "max", x_scale],
+                out=nn.Tensor.placeholder(x_scale.shape, x_scale.dtype),
+            )
+            x_q = (x / x_scale).astype(self.config.activation_dtype)
+        else:
+            raise ValueError(f"Unknown calibration mode: {self.config.calibration_mode}")
+
+        if self.config.weight_dtype == self.config.storage_dtype and not self.config.use_scale:
             w = self.q_weight
+            w = nn.op.permute_dims(w)
+            x = nn.op.matmul(
+                x_q, w, out_dtype=self.out_dtype
+            )  # mixed precision matmul: fp8 * fp8 => fp16
         else:
+            # dequantize input and weight to fp16, this can be fused into matmul during lowering
+            x = nn.op.tensor_expr_op(
+                lambda quantized_x, scale: self.config._dequantize(  # pylint: disable=protected-access
+                    quantized_x, scale, out_shape=x.shape
+                ),
+                "dequantize_x",
+                args=[x_q, x_scale],
+            )
```

Comment:
Just want to better understand: where and how was `x` quantized before?

Reply:
before this PR we just converted to f8 via cast

Reply:
I see. So is the dequantization also just cast?

Reply:
right